text/0000-freeze.md

```diff
@@ -0,0 +1,155 @@
+
+The following example uses a `Mutex` (a `Send`/`Sync`, but not `Freeze` type to create UB):
+
+```rust
+let x = TCell::new(Mutex::new("hello there".to_owned()));
+
+// ..  inside a transaction
+let shallow_copy = x.borrow(tx, Default::default())?;
+// Locking a shallow copy of a lock... is not really a lock at all!
+// The original String is deallocated here, likely leading to double-frees.
+*shallow_copy.lock().unwrap() = "uh oh".to_owned();
+```
+
+By having snapshotting functions like `borrow` require `Freeze`, such disastrous situations are prevented at compile time, without being overly restrictive, or requiring slower heap allocation based workarounds.
+
+Similarly to the above example, `crossbeam` would be able to expand `Atomic` to include non-copy types. See [this](https://github.com/crossbeam-rs/crossbeam/issues/379) issue.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`Freeze` is a new marker trait, similar to `Send` and `Sync`, that is intended to only be implemented for types which have no direct interior mutability, and are therefore safe to place in read only memory.
+
+## What types are `Freeze`?
+
+The list of `Freeze` types is long, including primitives, `String`, `Vec`, `Option<String>`, `Box<T>`, `Arc<T>`, `Rc<T>`, etc. This is because you cannot modify the memory contained directly within these types through an immutable reference.
+
+Types that do not implement `Freeze` include types used in parallel programming such as, `Mutex<T>`, `AtomicUsize`, etc, as well as `Cell`, `RefCell`, and `UnsafeCell`. This is because their memory can be modified via an immutable reference.
+
+## My type doesn't implement `Freeze`, but I need it to be `Freeze`.
+
+To convert a type which is not `Freeze`, into a `Freeze` type, all that is required is to stick it on the heap. For example, `Box<T>` is `Freeze` even if `T` is an `UnsafeCell`.
+
+If you really know what you are doing, and promise not to mutate any data in your type through an immutable reference, then you can implement `Freeze` like so:
+
+```rust
+struct MyType { /* .. */ }
+unsafe impl Freeze for MyType {}
+```
+
+This requires `unsafe`, because UB is possible if in fact the memory occupied by `MyType` is mutable through an immutable reference to `MyType`.
```

Comment:
I think this section should go into much more detail about what exactly constitutes undefined behavior. cc @RalfJung

Reply:
Hm... the impact of `UnsafeCell` on UB currently is that it is the only way writing to a ptr derived from a shared reference can ever be allowed. https://github.com/rust-lang/unsafe-code-guidelines/issues/236 describes in detail how that works. This RFC *at least* change the fact that checking for `Freeze` is "purely an optimization", and as such it should describe in detail what the new rules are.

Reply:
The RFC should also describe how the Language Reference should be updated, which [currently says that this is UB](https://doc.rust-lang.org/reference/behavior-considered-undefined.html):

> Mutating immutable data. All data inside a const item is immutable. Moreover, all data reached through a shared reference or data owned by an immutable binding is immutable, unless that data is contained within an UnsafeCell<U>.

Reply:
Resolved by forbidding `unsafe impl Freeze`.